test/linalg/symmetric.jl

```diff
@@ -42,208 +42,265 @@ end
+                        eig(Symmetric(asym), d[1] - 1, (d[2] + d[3])/2) # same result, but checks that method works
+                        @test eigvals(Symmetric(asym), 1:2) ≈ d[1:2]
+                        @test eigvals(Symmetric(asym), d[1] - 1, (d[2] + d[3])/2) ≈ d[1:2]
+                        # eigfact doesn't support Symmetric{Complex}
+                        @test full(eigfact(asym)) ≈ asym
+                        @test eigvecs(Symmetric(asym)) ≈ eigvecs(asym)
+                    end
+
+                    d, v = eig(aherm)
+                    @test aherm*v[:,1] ≈ d[1]*v[:,1]
+                    @test v*Diagonal(d)*v' ≈ aherm
+                    @test isequal(eigvals(aherm[1]), eigvals(aherm[1:1,1:1]))
+                    @test abs.(eigfact(Hermitian(aherm), 1:2)[:vectors]'v[:,1:2]) ≈ eye(eltya, 2)
+                    eig(Hermitian(aherm), 1:2) # same result, but checks that method works
+                    @test abs.(eigfact(Hermitian(aherm), d[1] - 1, (d[2] + d[3])/2)[:vectors]'v[:,1:2]) ≈ eye(eltya, 2)
+                    eig(Hermitian(aherm), d[1] - 1, (d[2] + d[3])/2) # same result, but checks that method works
+                    @test eigvals(Hermitian(aherm), 1:2) ≈ d[1:2]
+                    @test eigvals(Hermitian(aherm), d[1] - 1, (d[2] + d[3])/2) ≈ d[1:2]
+                    @test full(eigfact(aherm)) ≈ aherm
+                    @test eigvecs(Hermitian(aherm)) ≈ eigvecs(aherm)
+
+                    # relation to svdvals
+                    if eltya <: Real #svdvals! has no method for Symmetric{Complex}
+                        @test sum(sort(abs.(eigvals(Symmetric(asym))))) == sum(sort(svdvals(Symmetric(asym))))
+                    end
+                    @test sum(sort(abs.(eigvals(Hermitian(aherm))))) == sum(sort(svdvals(Hermitian(aherm))))
+                end
+
+                @testset "det" begin
+                    @test det(aherm) ≈ det(Hermitian(aherm, :U))
+                    @test det(aherm) ≈ det(Hermitian(aherm, :L))
+                    @test det(asym)  ≈ det(Symmetric(asym, :U))
+                    @test det(asym)  ≈ det(Symmetric(asym, :L))
+                end
+
+                @testset "rank" begin
+                    let A = a[:,1:5]*a[:,1:5]'
+                        # Make sure A is Hermitian even in the presence of rounding error
+                        # xianyi/OpenBLAS#729
+                        A = (A' + A) / 2
+                        @test rank(A) == rank(Hermitian(A))
+                    end
+                end
+
+                @testset "pow" begin
+                    @test (asym)^2     ≈ Array(Symmetric(asym)^2)
+                    @test (asym)^-2    ≈ Array(Symmetric(asym)^-2)
+                    @test (aherm)^2    ≈ Array(Hermitian(aherm)^2)
+                    @test (aherm)^-2   ≈ Array(Hermitian(aherm)^-2)
+                    if eltya == Int
+                        @test (asym)^2.0   ≈ real(Array(Symmetric(asym)^2.0))
```

Comment:
@tkelman if the input matrix is diagonal, then the power to a real number method returns a matrix of real numbers. If not, it may return an array of complex numbers. The output type isn't fixed.